crates/sui-core/src/checkpoints/mod.rs

```diff
@@ -737,6 +738,107 @@ impl CheckpointStore {
+    ) {
+        info!("rexecuting locally computed checkpoints for crash recovery");
+        let epoch = epoch_store.epoch();
+        let highest_executed = self
+            .get_highest_executed_checkpoint_seq_number()
+            .expect("get_highest_executed_checkpoint_seq_number should not fail")
+            .unwrap_or(0);
+
+        let Some(highest_built) = self.get_latest_locally_computed_checkpoint() else {
+            info!("no locally built checkpoints to verify");
+            return;
+        };
+
+        for seq in highest_executed + 1..=*highest_built.sequence_number() {
+            info!(?seq, "Re-executing locally computed checkpoint");
+            let Some(checkpoint) = self
+                .get_locally_computed_checkpoint(seq)
+                .expect("get_locally_computed_checkpoint should not fail")
+            else {
+                panic!("locally computed checkpoint {:?} not found", seq);
+            };
+
+            let Some(contents) = self
+                .get_checkpoint_contents(&checkpoint.content_digest)
+                .expect("get_checkpoint_contents should not fail")
+            else {
+                panic!("checkpoint contents not found for locally computed checkpoint {:?} (digest: {:?})", seq, checkpoint.content_digest);
+            };
+
+            let cache = state.get_cache_reader();
+
+            let tx_digests: Vec<_> = contents.iter().map(|digests| digests.transaction).collect();
+            let fx_digests: Vec<_> = contents.iter().map(|digests| digests.effects).collect();
+            let txns = cache
+                .multi_get_transaction_blocks(&tx_digests)
+                .expect("multi_get_transaction_blocks should not fail");
+            for (tx, digest) in txns.iter().zip(tx_digests.iter()) {
+                if tx.is_none() {
+                    panic!("transaction {:?} not found", digest);
+                }
+            }
+
+            let txns: Vec<_> = txns
+                .into_iter()
+                .map(|tx| tx.unwrap())
+                // end of epoch transaction can only be executed by CheckpointExecutor
+                .filter(|tx| !tx.data().transaction_data().is_end_of_epoch_tx())
+                .zip(fx_digests.into_iter())
+                .map(|(tx, fx)| {
+                    (
+                        VerifiedExecutableTransaction::new_from_checkpoint(
+                            (*tx).clone(),
+                            epoch,
+                            seq,
+                        ),
+                        fx,
+                    )
+                })
+                .collect();
+
+            info!(
+                ?seq,
+                ?tx_digests,
+                "Re-executing transactions for locally built checkpoint"
+            );
+            // this will panic if any re-execution diverges from the previously recorded effects digest
+            state.enqueue_with_expected_effects_digest(txns, epoch_store);
```

Comment:
Would we get better throughput if we schedule all transactions altogether instead of checkpoint by checkpoint?

Reply:
maybe, but it shouldn't matter - this would typically only be a a few seconds worth of checkpoints, and this code only ever runs after a restart